Enable import-module to be case insensitive

SteveL-MSFT opened the pull request:
Apparently macOS is case insensitive

Fix https://github.com/PowerShell/PowerShell/issues/1621
<!--

If you are a PowerShell Team member, please make sure you choose the Reviewer(s) and Assignee for your PR.
If you are not from the PowerShell Team, you can leave the fields blank and the Maintainers will choose them for you. If you are familiar with the team, feel free to mention some Reviewers yourself.

For more information about the roles of Reviewer and Assignee, refer to [CONTRIBUTING.md](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md).

-->

SteveL-MSFT committed:
enable import-module to be case insensitive

iSazonov commented on src/System.Management.Automation/engine/Modules/ModuleCmdletBase.cs:
```diff
@@ -323,6 +323,46 @@ internal List<WildcardPattern> MatchAll
                 }
             }
 
+            if (!found)
```
If it is only for Unix maybe use "#if UNIX"?

SteveL-MSFT replied:
it's technically any filesystem that's case sensitive and technically (although unlikely), it seems you can make NTFS case sensitive.  for practical cases, it's probably fine to #ifdef this and if a customer has an issue on Windows, we can revisit

SteveL-MSFT replied:
fixed and added comment

adityapatwardhan commented on src/System.Management.Automation/engine/Modules/ModuleCmdletBase.cs:
```diff
@@ -323,6 +323,46 @@ internal List<WildcardPattern> MatchAll
                 }
             }
 
+            if (!found)
```
The cyclomatic complexity of this function is pretty high. Can this be refactored into a smaller function?

SteveL-MSFT replied:
I could simplify it a bit if we can make the assumption that the psd1 or psm1 file always has the same casing as the folder name.  It's not clear to me if that is a safe assumption or not.  The current code is conservative in that you could have: "...\Modules\FOO\foo.psd1"

SteveL-MSFT replied:
I'll check with the PSGet folks

SteveL-MSFT replied:
per PSGet, it is safe to assume the folder and file names have the same casing.  So I'll simplify this code.

adityapatwardhan commented on test/powershell/Modules/Microsoft.PowerShell.Core/Import-Module.Tests.ps1:
```diff
@@ -124,3 +124,56 @@ namespace ModuleCmdlets
+        @{modulePath="TESTMODULE"    ; manifest=$false}
+        ) {
+        param ($modulePath, $manifest)
+        New-Item -ItemType Directory -Path "$modulesPath/$modulePath" -Force
```
Please redirect to null.
New-Item -ItemType Directory -Path "$modulesPath/$modulePath" -Force > $null

SteveL-MSFT replied:
will add

adityapatwardhan commented on test/powershell/Modules/Microsoft.PowerShell.Core/Import-Module.Tests.ps1:
```diff
@@ -124,3 +124,56 @@ namespace ModuleCmdlets
+            New-ModuleManifest -Path "$modulesPath/$modulePath/TESTMODULE.psd1" -RootModule "TESTMODULE.psm1" -ModuleVersion 1.1
+        }
+        Set-Content -Path "$modulesPath/$modulePath/TESTMODULE.psm1" -Value "function mytest { 'hello' }"
+        Import-Module testMODULE
```
Maybe `-Force` to ensure you are not loading a cached version.

SteveL-MSFT replied:
will add

SteveL-MSFT replied:
actually, as part of this test, I'm removing the module with a different casing, so I don't want to -force loading

adityapatwardhan replied:
Ok

adityapatwardhan commented on test/powershell/Modules/Microsoft.PowerShell.Core/Import-Module.Tests.ps1:
```diff
@@ -124,3 +124,56 @@ namespace ModuleCmdlets
+    }
+
+    It "Import-Module will import exact casing if available" {
+        New-Item -ItemType Directory -Path "$modulesPath\Test" -Force
```
`> $null`

SteveL-MSFT replied:
will add

adityapatwardhan commented on test/powershell/Modules/Microsoft.PowerShell.Core/Import-Module.Tests.ps1:
```diff
@@ -124,3 +124,56 @@ namespace ModuleCmdlets
+    It "Import-Module will import exact casing if available" {
+        New-Item -ItemType Directory -Path "$modulesPath\Test" -Force
+        Set-Content -Path "$modulesPath\Test\Test.psm1" -Value "function casetest { 'first' }"
+        New-Item -ItemType Directory -Path "$modulesPath\tEST" -Force
```
`> $null`

SteveL-MSFT replied:
will add

SteveL-MSFT committed:
address PR feedback
simplify search loops with assumption that module file casing matches module folder name which
is appropriate per PSGet

SteveL-MSFT commented:
@adityapatwardhan feedback addressed

iSazonov commented on src/System.Management.Automation/engine/Modules/ModuleCmdletBase.cs:
```diff
@@ -323,6 +323,38 @@ internal List<WildcardPattern> MatchAll
+            {
+                // try a case-insensitive search for the module folder
+                // based on PSGet, we can assume the module file casing matches the folder
+                foreach (string path in modulePath)
```
In line 294 the same cycle? Can we remove it and replace with the new one on all platforms?

SteveL-MSFT replied:
We still need both as the first one will find an exact match and fall through to this if it doesn't find it.  Also the above loop also handles a file path to the module which I think should be case sensitive since it's a literal file path.  The lower code only handles the module name which we can treat as case insensitive.  The code is similar, but I don't think it's worth combining which will make it more complex.

iSazonov replied:
Why we should use different logic for module folder names and module file names?

SteveL-MSFT replied:
If someone uses: `Import-module /users/steve/dev/MyModule/MyModule.psd1` then that should be treated as a case-sensitive path.  If they use `Import-module mymodule`, then that's case-insensitive as that's a module name and not a filepath.

iSazonov replied:
Ah, I meant folder and file names not paths.

SteveL-MSFT replied:
I don't see an easy way to have a single loop between the two without incurring a cost of enumerating the folders for the purpose of case-insensitive comparison in the "normal" case where either the case matches or the filesystem is case-insensitive.  If there's two modules "Foo" and "foo", and the user does `import-module foo` we want the first loop to match `foo` and not `Foo` as that is an exact match.

iSazonov replied:
We have the same problem in #4699 - can you review the idea to use here?

SteveL-MSFT replied:
In the tab completion case, I think you have to cover more cases as `ab` might match `ABc`, `abc`, or `aBcdefg`, etc...  so creating a case-insenstive match pattern makes sense.  In this case, we have a specific name (wildcard is not accepted) and just need a case-insensitive match if we haven't already matched the exact casing (which is preferred).  And we don't need to return all matches, the first one wins.

If we want just the first match to win even if there is an exact match, it would be easy to merge the two loops.  However, I think we want to get an exact match if it exists, then fallback on case-sensitive search.  For example, if the user typed `ipmo AAA` which exists in the last modulepath, but `aaa` exists in the first, I would have to enumerate all module folder names in every modulepath part every time just to make sure `AAA` wasn't found after my first case-insensitive match.

iSazonov replied:
We could have solved this situation on the other hand. Do we want _all_ the modules to be ported? If so, we must document this and exclude the case-sensitivity in this code.

SteveL-MSFT replied:
That's a good point on portable scripts being able to rely on module names being case insensitive.  It would make sense to just say module names are case insensitive everywhere and have it enforced by install-module and publish-module.

SteveL-MSFT committed:
[feature]
combine loops

iSazonov commented on src/System.Management.Automation/engine/Modules/ModuleCmdletBase.cs:
```diff
@@ -293,34 +293,48 @@ internal List<WildcardPattern> MatchAll
+                                qualifiedPath = Path.Combine(folder, fileName);
+#else
+                                qualifiedPath = Path.Combine(qualifiedPath, fileBaseName);
+#endif
```
Perhaps we could exclude `#if` here and leave one line.

SteveL-MSFT replied:
I think this line isn't needed as I have qualifiedPath = folder above, was probably thinking this, but didn't put it together.  Will fix.

SteveL-MSFT committed:
[feature]
removed test that is no longer valid

iSazonov commented:
Should we add negative tests too (to reflect the situations we have decided to exclude)?  What errors should appear?
Also why we skip Windows in new test? If we want to have all modules portable all test should run on all platforms identically.

SteveL-MSFT commented:
@iSazonov test was originally skipped due to the previous code for an exact match since it wasn't possible to create two modules that differ only by casing on Windows.  With the new change, no reason not to run it on Windows.  I also noticed a bug in the test where I wasn't removing the test module between runs.

SteveL-MSFT committed:
[feature]
enable test for Windows, removed invalid test where a module with version folder requires a manifest

iSazonov commented:
I don't see still that tests reflect to "StringComparison.OrdinalIgnoreCase".

SteveL-MSFT commented:
@iSazonov the new test creates a module called TESTMODULE and imports it using testMODULE, this is where the case-insensitive comparison occurs.  In the UNIX code path, if it enumerates the module folders in the module path and sees if it matches the name of the given module.  Not clear to me if you were expecting a different test?

iSazonov commented:
@SteveL-MSFT My thoughts was that we should do `Import-Module testMODULE` and `Import-Module TESTMODULE` and check that we load the same module. Although I'm not sure it's necessary.

SteveL-MSFT commented:
@iSazonov I don't think that's necessary

adityapatwardhan reviewed (approved):
LGTM

adityapatwardhan commented on src/System.Management.Automation/engine/Modules/ModuleCmdletBase.cs:
```diff
@@ -293,34 +293,42 @@ internal List<WildcardPattern> MatchAll
+                            {
+                                qualifiedPath = Path.Combine(qualifiedPath, fileBaseName);
+                            }
+                            else if (Directory.Exists(qualifiedPath))
```
Please use Utils.NativeDirectoryExists()
https://github.com/PowerShell/PowerShell/blob/master/src/System.Management.Automation/engine/Utils.cs#L912

SteveL-MSFT replied:
Will change

adityapatwardhan commented on src/System.Management.Automation/engine/Modules/ModuleCmdletBase.cs:
```diff
@@ -293,34 +293,42 @@ internal List<WildcardPattern> MatchAll
-
-                if (!found)
+#if UNIX
+                foreach (string folder in Directory.EnumerateDirectories(path))
```
Please use Utils.NativeEnumerateDirectory()
https://github.com/PowerShell/PowerShell/blob/master/src/System.Management.Automation/engine/Utils.cs#L926

SteveL-MSFT replied:
Will change

SteveL-MSFT replied:
This method appears to be only implemented for Win32.  Since this path is only for UNIX, I'll leave it as-is.

adityapatwardhan replied:
ok

SteveL-MSFT committed:
[feature]
changed to using Utils.NativeDirectoryExists() for perf reasons
